buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.gradle.plugin;
+
+import org.elasticsearch.gradle.VersionProperties;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.InvalidUserDataException;
+import org.gradle.api.Project;
+import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Creates a plugin descriptor.
+ */
+public class PluginPropertiesTask extends DefaultTask {
+
+    private PluginPropertiesExtension extension;
+
+    private File descriptorOutput = new File(getProject().getBuildDir(),
+            "generated-resources/plugin-descriptor.properties");
+
+    public PluginPropertiesTask() {
+        Project project = this.getProject();
+        this.extension = project.getExtensions().create("esplugin", PluginPropertiesExtension.class, project);
```

Comment:
sorry I missed this in my previous review,
Creating the extension here  is part of the configuration we want to avoid doing as part of the constructor.
The proper place for it is in the `PluginBuilderPlugin`. We can then reference it here ( using get by name instead of create ) and we must also declare it as an `@Input` as we want the task to be re-run if any of it changes.

Reply:
Can you help me with understanding what is `PluginBuilderPlugin` I could not find a reference to it anywhere? :confused:

Reply:
Sorry, I made a typo in the name the class is called `PluginBuildPlugin`

Reply:
I might need help with this one. I don't know how to modify the `PluginBuildPlugin.groovy` and I tried adding the the `plugin` as follow (using create, add, etc.)  

```project.extensions.create('esplugin', PluginPropertiesExtension, project)``` 

And seems like it adds it but, then in the `PluginPropertiesTask` inside the constructor 

```this.extension = project.getExtensions().getByType(PluginPropertiesExtension.class);``` 
and 
```this.extension = (PluginPropertiesExtension) project.getExtensions().getByName('esplugin');```

Fails complaining that it cannot find the plugin. 

```
Caused by: org.gradle.api.UnknownDomainObjectException: Extension of type 'PluginPropertiesExtension' does not exist. Currently registered extension types: [ExtraPropertiesExtension, DefaultArtifactPublicationSet, ReportingExtension, SourceSetContainer, JavaPluginExtension]
``` 
The above is when I try to retrieve the plugin via `getByType` in the `PluginPropertiesTask` and when I try to retrieve it with `getByName` then error is as follow 

```
org.gradle.api.UnknownDomainObjectException: Extension with name 'esplugin' does not exist. Currently registered extension names: [ext, defaultArtifacts, reporting, sourceSets, java]
```
It is almost like `accessible` in `PluginBuildPlugin` (because when I try it again, it complains that it's already registered) but, not accessible in `PluginPropertiesTask`.

Reply:
Make sure that the extension is registered _before_ the task is created. 
That seems obvious but sometimes it's easy to overlook ordering with Gradle. 
Other than that what you are doing looks right.
Tasks also have extensions, in a different namespace, but you are correctly trying to look it up from the project.

Reply:
@rhamedy thanks for continuing to pursue this!  If you can't get it to work feel free to push the code and I'll take a look.

Reply:
I have done the above refactoring! Please see the recent changes.